### PR TITLE
Switch router store to redux

### PR DESCRIPTION
**Pull request by aenario**
For review, please have a look at these somewhat more readable PRs : 
https://github.com/aenario/cozy-emails/pull/1 
https://github.com/aenario/cozy-emails/pull/2 
https://github.com/aenario/cozy-emails/pull/3 
https://github.com/aenario/cozy-emails/pull/4 
https://github.com/aenario/cozy-emails/pull/5 
https://github.com/aenario/cozy-emails/pull/6 

Not included here, will make a separate PR afterward
https://github.com/aenario/cozy-emails/pull/7 

# What was done

This huge PR complete the transition to Redux with the following parts : 

- Make the state fully immutable, and start to use Immutable.Record model for schema enforcement and to attach function to prototype (initially to remove these functions that dont use the state from getters) 

- Transform router_store, the core of the application into several reducers :
  - reducers/route handle the current url
  - reducers/modal handle the modal
  - reducers/message_fetch handle the state of fetching messages to fill the current view.

- Following what was done in previous PR, transform all getters into pure getters, ie. function(state, args ...). These pure getters are placed in the puregetters until the transition is complete. Then renamed in https://github.com/aenario/cozy-emails/pull/7

- Also transform ActionCreators in pure function(dispatch), this change is incomplete and need to be discussed cf difference between MessageActionCreator and RouterActionCreator

- Following these two changes, we can use `react-redux/connect` to pass all data and actions as props from the container component. For now only `components/application` is a container component, we need to progressively switch all panel (message-list, conversation) to container as well.

- The Router has been simplified and divided between the application specific code `client/app/routes.coffee` which contains routes and reverse routing and can be used from reducers & getters without introducing backbone router dependency and `client/app/router.coffee` which is the Backbone.Router with 

- Some linting were applied in specs.js files and all around

- Some utility function were moved from getters to components utils, they will then be moved to libs in https://github.com/aenario/cozy-emails/pull/7 

# "Golden rules" of the new structure

- A reducer is a function(state,action) -> newState (redux)
    - newState should always be !== state (immutable)
- A getter is a function(wholeAppState) -> transformed part of the state
- A getter should do one thing and only thing, ie split getMailboxById from getCurrentMailbox

# What to do

- create conversation_panel / list_panel / menu components to be container
- Agree on proper ActionCreator structure and progressively unify toward it.
- Use `reselect` or alternative to memoize getters and increase performance WHEN NECESSARY

### CODE REVIEW

File: client/test/router_store.spec.js [hunk 1]
```diff
@@ -63,18 +70,17 @@ describe('RouterStore', () => {
   });
 
   beforeEach(() => {
-    Dispatcher.dispatch({
-      type: ActionTypes.ROUTES_INITIALIZE,
-      value: router,
-    });
+    Dispatcher.dispatch({type: ActionTypes.RESET_FOR_TESTS});
```
Comment (misstick) [inline client/test/router_store.spec.js]: @aenario : Why not keep this event name? What the issue about this? I think it's important not to create specific workflow for test but keep real one to make test as relevant as possible.

File: client/test/message_store.spec.js [hunk 2]
```diff
@@ -110,7 +110,7 @@ describe('Message Store', () => {
       });
     });
 
-    it('isUnread', () => {
+    it.skip('isUnread', () => {
```
Comment (misstick) [inline client/test/message_store.spec.js]: @aenario : Why have you commented this? Could you had a comment to explain context please? (not specified into commit description)

File: client/test/message_store.spec.js [hunk 3]
```diff
@@ -121,7 +121,7 @@ describe('Message Store', () => {
       assert.isFalse(MessageStore.isUnread({ message: input }));
     });
 
-    it('isFlagged', () => {
+    it.skip('isFlagged', () => {
```
Comment (misstick) [inline client/test/message_store.spec.js]: @aenario : same as above :)

File: client/test/message_store.spec.js [hunk 4]
```diff
@@ -132,7 +132,7 @@ describe('Message Store', () => {
       assert.isFalse(MessageStore.isFlagged({ message: input }));
     });
 
-    it('isAttached', () => {
+    it.skip('isAttached', () => {
```
Comment (misstick) [inline client/test/message_store.spec.js]: @aenario : same as above :)

File: client/test/router_action_creator.spec.js [hunk 5]
```diff
@@ -10,7 +10,7 @@ const SpecDispatcher = require('./utils/specs_dispatcher');
 const ActionTypes = require('../app/constants/app_constants').ActionTypes;
 
 
-describe('RouterActionCreator', () => {
+describe.skip('RouterActionCreator', () => {
```
Comment (misstick) [inline client/test/router_action_creator.spec.js]: @aenario : same as above :)

File: client/test/router_store.spec.js [hunk 6]
```diff
@@ -103,7 +109,7 @@ describe('RouterStore', () => {
     });
 
 
-    it('getRouter', () => {
+    it.skip('getRouter', () => {
```
Comment (misstick) [inline client/test/router_store.spec.js]: @aenario : same as above :)

File: client/test/router_store.spec.js [hunk 7]
```diff
@@ -122,7 +128,7 @@ describe('RouterStore', () => {
         assert.equal(RouterStore.getAccountID(), undefined);
       });
 
-      it('Should save `value`', () => {
+      it.skip('Should save `value`', () => {
```
Comment (misstick) [inline client/test/router_store.spec.js]: @aenario : same as above :)

**Comment by misstick** [comment]
@aenario can you add a description and flags to this PR to understand scope and status of this PR please?

File: client/app/constants/app_constants.coffee [hunk 8]
```diff
@@ -135,6 +135,8 @@ module.exports =
         'VIEW_ACTION'   : 'VIEW_ACTION'
         'SERVER_ACTION' : 'SERVER_ACTION'
 
+    DefaultActions:
```
Comment (misstick) [inline client/app/constants/app_constants.coffee]: 👍

Comment (misstick) [inline client/app/constants/app_constants.coffee]: 👍

File: client/test/account_store.spec.js [hunk 9]
```diff
@@ -396,8 +396,26 @@ describe('AccountStore', () => {
 
 
     describe('Should REMOVE account(s)', () => {
-      it.skip('REMOVE_ACCOUNT_SUCCESS', () => {
-        // FIXME Nothing is done here in the store code.
+      beforeEach(() => {
```
Comment (misstick) [inline client/test/account_store.spec.js]: @aenario : shouldn't you reset accounts after each test? ie. https://github.com/aenario/cozy-emails/blob/41b0c037ddbdeb44c419ad039b0118cfade4827a/client/test/account_store.spec.js#L302

File: client/app/puregetters/account.coffee [hunk 10]
```diff
@@ -26,7 +26,11 @@ module.exports =
         @getAllAccounts(state).find (account) ->
             # TODO There are some cases where account.get is not a function
             # this must be investigated
-            return account.get?('mailboxes').get(mailboxID)
+            unless account.get
```
Comment (misstick) [inline client/app/puregetters/account.coffee]: 👍 

@aenario : We didn't handle failures until now; will this be the new pattern?

File: client/test/router_store.spec.js [hunk 11]
```diff
@@ -1066,15 +1060,12 @@ describe('RouterStore', () => {
     beforeEach(() => {
       resetStore();
     });
-    // Skipped after account redux migration
-    // TODO: make the acount fixture as Immutable
     describe('getMessagesPerPage', () => {
+      // ROMAINEDIT : this should always default to default value
```
Comment (misstick) [inline client/test/router_store.spec.js]: @aenario I'm not sure that this thought is very understandable ;)

File: client/test/utils/specs_dispatcher.js [hunk 12]
```diff
@@ -14,13 +14,8 @@ function freshRequire(path) {
   return require(path); // eslint-disable-line global-require
 }
 
-let lastStore = null;
-
 module.exports = () => {
   const store = freshRequire('../../app/reducers/_store');
-  if (lastStore && lastStore === store)
```
Comment (misstick) [inline client/test/utils/specs_dispatcher.js]: @aenario : why did you remove this?

File: client/test/utils/pseudorandom.js [hunk 13]
```diff
@@ -1,7 +1,11 @@
 'use strict';
 
 
-let seed = 0;
+let seed = process.env.PSEUDORANDOMSEED === undefined ? Math.random() :
+    parseInt(process.env.PSEUDORANDOMSEED, 10);
+
+// eslint-disable-next-line no-console
+console.log('SEED is', seed);
```
Comment (misstick) [inline client/test/utils/pseudorandom.js]: @aenario: is this resolve issues about random failure into tests?

File: client/app/constants/app_constants.coffee [hunk 14]
```diff
@@ -92,8 +96,7 @@ module.exports =
         'SETTINGS_UPDATE_FAILURE'   : 'SETTINGS_UPDATE_FAILURE'
 
         # Tasks
-        'RECEIVE_TASK_UPDATE'       : 'RECEIVE_TASK_UPDATE'
-        'RECEIVE_TASK_DELETE'       : 'RECEIVE_TASK_DELETE'
+        'CLICKED_TASK_OK'           : 'CLICKED_TASK_OK'
```
Comment (misstick) [inline client/app/constants/app_constants.coffee]: What does that mean? I'm not sure that using `clicked` is a good idea: it's about how the event was called not what should be done next. Shouldnt all event name follow this 2nd pattern?

File: client/app/constants/app_constants.coffee [hunk 15]
```diff
@@ -111,6 +114,8 @@ module.exports =
         'TOASTS_SHOW'               : 'TOASTS_SHOW'
         'TOASTS_HIDE'               : 'TOASTS_HIDE'
 
+        # Test
+        'RESET_FOR_TESTS'               : 'RESET_FOR_TESTS'
```
Comment (misstick) [inline client/app/constants/app_constants.coffee]: I'm not sure that creating specific events for test is a good practice; Does any other event able to replace it?

File: client/app/puregetters/requests.coffee [hunk 16]
```diff
@@ -0,0 +1,106 @@
...
+        state.get('refreshes')
+
+    isRefreshError: (state) ->
+        @getRefreshes(state).get('errors')?.length
+
+    isRefreshing: (state)->
+        @getRefreshes(state).size isnt 0 or
+        @isLoading(state, Requests.REFRESH_MAILBOX)
+
+    isIndexing: (state, accountID) ->
```
Comment (misstick) [inline client/app/puregetters/requests.coffee]: Is this still relevant? Why comment all this method?
Can you add some comments to explain all this (according to our discussions during poker planning) please?

I think there is a misunderstanding into the implementation (because of my work mainly, sorry) between : import and indexing. Add comment may be a solution to explain that and how to change it.

File: client/app/reducers/messagefetch.coffee [hunk 17]
```diff
@@ -48,6 +48,7 @@ module.exports = (state = DEFAULT_STATE, action, appstate) ->
             state = setOnRequest(state, URIKey, 'isLoading', false)
 
             pageAfter = result.messages.reduce(findOldest, null)?.date
+            console.log('NOW', URIKey, pageAfter)
```
Comment (m4dz) [inline client/app/reducers/messagefetch.coffee]: `console` oops :P

File: client/app/actions/message_action_creator.coffee [hunk 18]
```diff
@@ -1,7 +1,7 @@
 {ActionTypes}   = require '../constants/app_constants'
 XHRUtils        = require '../libs/xhr'
 
-MessageGetters  = require '../puregetters/message'
+MessageGetters  = require '../puregetters/messages'
```
Comment (misstick) [inline client/app/actions/message_action_creator.coffee]: Could you add into this PR's description le list of the functional scenari you fixed (such as this one) plz?

File: client/app/constants/app_constants.coffee (hunk 14 above)
Comment (aenario) [inline client/app/constants/app_constants.coffee]: Actions should be event unless we opt for using saga. To be discussed.

File: client/test/utils/pseudorandom.js (hunk 13 above)
Comment (aenario) [inline client/test/utils/pseudorandom.js]: No, but it should allow us to debug when they happen. Ie "tests fail when SEED IS ...."

File: client/app/actions/message_action_creator.coffee (hunk 18 above)
Comment (aenario) [inline client/app/actions/message_action_creator.coffee]: Mmmh, to be honnest, my tests so far are more of the randomly click everywhere and fix the errors spawning in browser console. I will try to make a list

File: client/app/puregetters/requests.coffee (hunk 16 above)
Comment (aenario) [inline client/app/puregetters/requests.coffee]: Not sure if these are still relevant. These is mostly translated code from previous state. Let's agree on naming, events and expected behaviours tomorrow and then we can remove what will not be used.

File: client/test/account_store.spec.js (hunk 9 above)
Comment (aenario) [inline client/test/account_store.spec.js]: RESET_ACCOUNT is a test-specific action and should be removed. I think the proper way to test is to start from a clean state (it's what the freshRequire(store) in specs_dispatcher is for). 
Or to test reducers as unit.
To be discussed.

File: client/test/router_action_creator.spec.js (hunk 5 above)
Comment (aenario) [inline client/test/router_action_creator.spec.js]: Mmmh, actually I am not sure, but anyway the tests inside are empty.

File: client/test/message_store.spec.js (hunk 2 above)
Comment (aenario) [inline client/test/message_store.spec.js]: I will add comment to every skipped test.

File: server/models/mailbox.coffee [hunk 19]
```diff
@@ -63,18 +63,6 @@ class Mailbox extends cozydb.CozyModel
         else if /all/i.test path
             return 'allMailbox'
 
-
```
Comment (misstick) [inline server/models/mailbox.coffee]: Why pushing this if you are not sure of the result?

File: client/test/router_action_creator.spec.js (hunk 5 above)
Comment (misstick) [inline client/test/router_action_creator.spec.js]: Maybe its missing comments : some tests have been written and added to the file as empty one to be filled in time. Maybe add a comment : "should be written" or maybe someting better describing why they are empty?

File: server/models/mailbox.coffee (hunk 19 above)
Comment (aenario) [inline server/models/mailbox.coffee]: It's the removed code that should never have been commited.

File: client/test/router_store.spec.js [hunk 20]
```diff
@@ -734,15 +732,15 @@ describe('RouterStore', () => {
...
       });
 
-      // Skipped after account redux migration
-      // TODO: make the acount fixture as Immutable
-      describe.skip('defaultView', () => {
-        it('Should goto `AccountNew` (no account found)', () => {
-          let url = RouterStore.getURL().replace('#', '');
+      describe('defaultView', () => {
+        //ROMAINEDIT : this is not handled at the getter level
+        it.skip('Should goto `AccountNew` (no account found)', () => {
```
Comment (misstick) [inline client/test/router_store.spec.js]: Should't you test `Route.makeURL` too?

File: client/test/router_store.spec.js [hunk 21]
```diff
@@ -734,15 +732,15 @@ describe('RouterStore', () => {
...
+        // Dispatcher.dispatch({ type: 'FORCE STORES TO GET INITIAL STATE' });
       });
 
-      // Skipped after account redux migration
-      // TODO: make the acount fixture as Immutable
-      describe.skip('defaultView', () => {
-        it('Should goto `AccountNew` (no account found)', () => {
-          let url = RouterStore.getURL().replace('#', '');
+      describe('defaultView', () => {
+        //ROMAINEDIT : this is not handled at the getter level
```
Comment (misstick) [inline client/test/router_store.spec.js]: Should't you test `Route.makeURL` too or instead? (and all other Router methods?)
This test is about how Router is handling redirection, shouldn't it be tested?

File: client/test/router_store.spec.js [hunk 22]
```diff
@@ -897,117 +889,59 @@ describe('RouterStore', () => {
     });
 
 
-    // TODO: testser avec options et parmas par défaut
-    describe('getCurrentURL', () => {
```
Comment (misstick) [inline client/test/router_store.spec.js]: Sorry but I missed that point; how is handled this method henceforth?

File: client/test/router_store.spec.js [hunk 23]
```diff
@@ -1266,15 +1183,15 @@ describe('RouterStore', () => {
...
+        });
 
         afterEach(() => {
           Dispatcher.dispatch({
             type: ActionTypes.MESSAGE_RESET_REQUEST,
           });
         });
 
-        it('With default MSGBYPAGE', () => {
+        it.skip('With default MSGBYPAGE', () => {
```
Comment (misstick) [inline client/test/router_store.spec.js]: Why is this skipped? Doesn't it works anymore?

File: client/test/router_store.spec.js [hunk 24]
```diff
@@ -1303,21 +1218,21 @@ describe('RouterStore', () => {
...
+          request = RouterStore.getLastFetch();
           request.page = 4;
           request.isComplete = true;
           request.hasNextPage = false;
           testMoreMessages(messages.slice(10, 20), request, true);
         });
 
 
-        it('With 5 MSGBYPAGE', () => {
+        it.skip('With 5 MSGBYPAGE', () => {
```
Comment (misstick) [inline client/test/router_store.spec.js]: same as above.

File: client/test/router_store.spec.js [hunk 25]
```diff
@@ -1460,29 +1367,44 @@ describe('RouterStore', () => {
...
 
       // No changes
       Dispatcher.dispatch({
         type: ActionTypes.REMOVE_ACCOUNT_SUCCESS,
-        value: { accountID: accountIDs.shift(), silent: true },
+        value: { accountID: firstAccount, silent: true },
       });
-      assert.equal(RouterStore.getAction(), null);
+
+      // ROMAINEDIT : action is never null
```
Comment (misstick) [inline client/test/router_store.spec.js]: Don't understand this comment :(
Any way the success/error redirection should be tested; it is failing (and wasn't weeks ago before Redux migration)

**Comment by misstick** [comment]
@aenario : When will you finish this? This PR was opened 24 days ago, it's a long time :)

**Comment by aenario** [comment]
Details of this PR : https://github.com/aenario/cozy-emails/pulls
